sycl/include/sycl/ext/oneapi/experimental/work_group_memory.hpp

```diff
@@ -0,0 +1,103 @@
+//===-------------------- work_group_memory.hpp ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <type_traits>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+template <typename T> struct is_unbounded_array : std::false_type {};
+
+template <typename T> struct is_unbounded_array<T[]> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_unbounded_array_v = is_unbounded_array<T>::value;
+
+class work_group_memory_impl {
+public:
+  work_group_memory_impl() : wgm_size{0}, buffer_size{0} {}
+  work_group_memory_impl(const work_group_memory_impl &rhs) = default;
+  work_group_memory_impl &
+  operator=(const work_group_memory_impl &rhs) = default;
+  work_group_memory_impl(size_t wgm_size, size_t buffer_size)
+      : wgm_size{wgm_size}, buffer_size{buffer_size} {}
+  size_t wgm_size;
+  size_t buffer_size;
```

Comment:
I haven't looked at this closely, but this looks suspicious.  The `work_group_memory` object is required to correspond to just a single Level Zero kernel parameter, which is a pointer to work-group-local memory.  This is the requirement specified here:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc#dpc-guaranteed-compatibility-with-level-zero-and-opencl-backends

It looks like `work_group_memory_impl` has two member variables of type `size_t`.  In addition, the `work_group_memory` type has a member variable of pointer type.  Does this mean that the Level Zero kernel will end up with three kernel parameters?

Reply:
Initial testing shows that the L0 kernel only ends up with one parameter, but I will make this into a proper test.

Reply:
I wonder why there is only one Level Zero parameter.  Is it because `wgm_size` and `buffer_size` are optimized away in the device code?  I'd be concerned about this because the requirement of one Level Zero parameter holds even when optimization is disabled.

When I look at the code, I don't see any uses of `buffer_size`.  Is that still needed?

If all we need is `wgm_size`, I wonder if it would be safer to implement `work_group_memory` such that its only data member was a union:

```
class work_group_memory {
 private:
  union {
    decoratedPtr ptr;  // Used only on device
    size_t size;       // Used only on host
  };
};
```

Or, you could use `std::variant` instead of a union in a similar way.

Reply:
So, my understanding is that the `processArg` function in `handler.cpp` file defines a mapping between the SYCL kernel parameters and the L0 kernel arguments. In the case of `work_group_memory`, I've defined the mapping to be such that whenever the runtime sees a work group memory parameter passed to a SYCL kernel, instead map that to a local memory buffer on the underlying backend where the size of the buffer is given by the buffer_size member of the work group memory object. If you look at my changes in `processArg` function, thats exactly what I'm doing. Therefore, with or without optimization, I believe there will only be one L0 kernel argument per work group memory object.

Reply:
The relevant lines are 793-798 in `handler.cpp` file

Reply:
OK, thanks for the explanation!

Reply:
@lbushi25 is right that this will only result in a single kernel argument, which is due to it being a "SYCL special class" so the arguments of its `__init` function is used as the arguments of the kernel and the function body is then used for constructing the object.

All that said however, I also have some concerns here:

1. These members are public and by public inheritance they will also be public in `work_group_memory`. They should not be, however.
2. Though this does not add extra arguments, they will increase register pressure (assuming the optimizer doesn't remove them.) We could move to a PIMPL implementation like many other SYCL classes (see for example `device`) where we have a `std::shared_ptr<work_group_memory_impl>` in this case, where `work_group_memory_impl` would be moved to a source file. Then, we could have something like:
```c++
#ifdef __SYCL_DEVICE_ONLY__
decoratedPtr ptr;
// To ensure we have the same object size on host and device, we add padding.
[[maybe_unused]] char padding[sizeof(std::shared_ptr<work_group_memory_impl>) - sizeof(decoratedPtr)];
#else
std::shared_ptr<work_group_memory_impl> impl;
#endif
```
Note that if we're sure we only ever need the two `size_t`, I don't think there's a big problem in having them directly in the class. Reading through it though, I don't fully understand why we need two `size_t`. If we could reduce it to one, we could do like Greg suggests further up.

Reply:
> These members are public and by public inheritance they will also be public in work_group_memory. They should not be, however. @AlexeySachkov corrected me and he's right. Since the inheritor is a class the default inheritance is private.

But I was also confused at first, so I think it worth explicitly declaring members of the base class as `protected`